WooCommerce/Classes/ViewRelated/Orders/Order Creation/ProductsSection/DiscountLineDetailsView.swift

```diff
@@ -10,72 +9,65 @@ struct DiscountLineDetailsView: View {
 
     var body: some View {
         VStack(alignment: .leading, spacing: .zero) {
-            Text(Localization.fixedPriceDiscountLabel)
-                .renderedIf(viewModel.feeOrDiscountType == .fixed)
-                .padding()
-            Text(Localization.percentagePriceDiscountLabel)
-                .renderedIf(viewModel.feeOrDiscountType == .percentage)
-                .padding()
-            HStack {
-                inputFixedField
-                    .renderedIf(viewModel.feeOrDiscountType == .fixed)
-                inputPercentageField
-                    .renderedIf(viewModel.feeOrDiscountType == .percentage)
-                HStack(spacing: 0) {
-                    Button(viewModel.currencySymbol) {
-                        viewModel.feeOrDiscountType = .fixed
-                    }
-                    .buttonStyle(isPrimary: viewModel.feeOrDiscountType == .fixed)
-                    .fixedSize(horizontal: true, vertical: false)
-                    Button(viewModel.percentSymbol) {
-                        viewModel.feeOrDiscountType = .percentage
-                    }
-                    .buttonStyle(isPrimary: viewModel.feeOrDiscountType == .percentage)
-                    .fixedSize(horizontal: true, vertical: false)
-                }
-                .padding(.trailing)
+            switch viewModel.feeOrDiscountType {
+            case .fixed:
+                LineDetailView(label: {
+                    Text(Localization.fixedPriceDiscountLabel)
+                }, content: {
+                    InputField(placeholder: Localization.fixedPriceDiscountInputPlaceholder,
+                               text: $viewModel.amount)
+                    discountTypeButtonToggle
+                        .padding()
+                })
+            case .percentage:
+                LineDetailView(label: {
+                    Text(Localization.percentagePriceDiscountLabel)
+                }, content: {
+                    InputField(placeholder: Localization.percentagePriceDiscountInputPlaceholder,
+                               text: $viewModel.percentage)
+                    discountTypeButtonToggle
+                        .padding()
+                })
             }
         }
     }
+}
 
-    private var inputFixedField: some View {
-        AdaptiveStack(horizontalAlignment: .leading) {
-            HStack {
-                BindableTextfield(Localization.fixedPriceDiscountInputPlaceholder,
-                                  text: $viewModel.amount,
-                                  focus: .constant(true))
-                .keyboardType(.numbersAndPunctuation)
+private extension DiscountLineDetailsView {
+    private var discountTypeButtonToggle: some View {
+        HStack(spacing: 0) {
+            Button(viewModel.currencySymbol) {
+                viewModel.feeOrDiscountType = .fixed
```

Comment:
❓ Not exactly part of the View changes in this PR but I just want to learn: Is this a common pattern in iOS to directly mutate a property in a viewmodel?

I'm more used to having a public function in the viewmodel to do this, something like `viewModel.updateDiscountType(.fixed)`

Reply:
That's a good question! Both approaches have pros and cons, since this is a simple Enum with only two mutually-exclusive cases and there's no additional or more complex logic to be performed it's fine to update the property directly, so we avoid another layer of indirection or extra complexity. Specially since the Enum is publicly accessible, having an update method or reaching directly to its cases has no difference than just switching its state.

For example, more related to activation you'd see a similar approach in `JetpackSetupViewModel`, when updating the `currentConnectionStep` since it's also a very simple Enum that just tracks current state where the only interested party is the view that owns that model.

I generally would add a public function like `updateDiscountType(FeeOrDiscountType)` if this becomes complex, a clear example is our Stores, for example [OrderAction](https://github.com/woocommerce/woocommerce-ios/blob/trunk/Yosemite/Yosemite/Actions/OrderAction.swift). There each case has several associated values, involves business logic, needs to be accessed from multiple places, needs to be testable, etc, ... so in this case makes sense to encapsulate these into methods and only expose what's necessary.